### PR TITLE
ws: forward top-level TLS options (rejectUnauthorized, ca, ...) to the connection

### PR DESCRIPTION
**Pull request by robobun**
### What

The npm `ws` package accepts TLS options as **top-level** options on the `WebSocket` client constructor and forwards them to the underlying `https.request`/`tls.connect`:

```js
new WebSocket("wss://localhost:8443", { rejectUnauthorized: false });
```

Bun's `ws` shim (`src/js/thirdparty/ws.js`) only read TLS options from `options.tls` (Bun's internal shape) or from an `agent`. It never read the top-level keys, so `rejectUnauthorized: false` was silently dropped and connecting to a self-signed `wss://` server failed with `TLS handshake failed`.

Fixes #31396.

### Reproduction

```js
import WebSocket from "ws";
// wss:// server with a self-signed cert
const ws = new WebSocket(`wss://localhost:${port}`, { rejectUnauthorized: false });
ws.on("open", () => console.log("open"));   // never fired
ws.on("error", err => console.error(err.message)); // "TLS handshake failed"
```

Node connects; Bun failed. Passing the same option via Bun's `tls: { rejectUnauthorized: false }` shape already worked — proving the native TLS layer was fine and the bug was purely option plumbing in the shim.

### Cause

In the constructor, TLS was read only from `options.tls`:

```js
tlsOptions = options?.tls;
```

Top-level `rejectUnauthorized`/`ca`/`cert`/`key`/… were never consulted.

### Fix

When no explicit `options.tls` object is given, collect the TLS keys the native WebSocket understands (the set `SSLConfig.fromJS` parses) off the top-level `options`:

```js
tlsOptions = $isObject(options.tls) ? options.tls : extractTlsOptions(options);
```

- An explicit `options.tls` stays authoritative, preserving Bun's existing shape.
- Booleans (`rejectUnauthorized`, …) are forwarded with a `!== undefined` check so an explicit `false` is honored.
- The agent path (`HttpsProxyAgent`, …) keeps its conservative key subset — agents put connection options in `connectOpts` that aren't all valid Bun TLS options (e.g. an array-form `ALPNProtocols`, which `SSLConfig.fromJS` rejects).

### Verification

New tests in `test/js/first_party/ws/ws-tls-options.test.ts` (alongside the existing `ws-proxy.test.ts`) stand up a self-signed `wss://` server via `Bun.serve({ tls })` and connect through the `ws` shim:

- `rejectUnauthorized: false` connects (fails before this change, passes after).
- Without the option, the self-signed server is still rejected (the option is actually doing something).
- An agent carrying `ca` plus a top-level `servername` connects — both TLS sources reach the handshake, covering the merge path.

The `ws` and `ws-proxy` suites are otherwise unchanged from their baseline.

### CODE REVIEW

File: src/js/thirdparty/ws.js [hunk 1]
```diff
@@ -158,7 +183,10 @@ class BunWebSocket extends EventEmitter {
     if ($isObject(options)) {
       headers = options?.headers;
       proxy = options?.proxy;
-      tlsOptions = options?.tls;
+      // `ws` forwards top-level TLS options (rejectUnauthorized, ca, cert, ...)
+      // to https.request/tls.connect. An explicit `tls` object stays
+      // authoritative for back-compat with Bun's existing shape.
+      tlsOptions = $isObject(options.tls) ? options.tls : extractTlsOptions(options);
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Good catch — fixed in c957b98. The agent fallback now merges instead of replacing (`tlsOptions = tlsOptions ? { ...agentOpts.tls, ...tlsOptions } : agentOpts.tls`), so top-level keys win on conflict while the agent fills gaps. Added a regression test (agent carries `ca`, top level passes `servername`, both must reach the handshake).

File: test/js/first_party/ws/ws-tls-options.test.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+//
+// The npm `ws` package accepts TLS options as top-level options on the
+// WebSocket constructor and forwards them to https.request/tls.connect:
+//
+//   new WebSocket("wss://host", { rejectUnauthorized: false });
+//
+// Bun's `ws` shim only read TLS options from `options.tls`, so top-level keys
+// like `rejectUnauthorized: false` were dropped and connecting to a self-signed
+// `wss://` server failed with "TLS handshake failed".
+describe("ws top-level TLS options", () => {
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/first_party/ws/ws-tls-options.test.ts:15]: Agreed — this is a never-supported feature, not a regression. Moved to `test/js/first_party/ws/ws-tls-options.test.ts` (next to the existing `ws-proxy.test.ts`) in ecc9ee5, and updated the PR description Verification section. Kept it as its own file rather than folding into `ws.test.ts` because that suite's subprocess echo-server tests time out under the debug/ASAN build and would muddy the regression signal.

File: src/js/thirdparty/ws.js [hunk 3]
```diff
@@ -16,46 +16,71 @@ const readyStates = ["CONNECTING", "OPEN", "CLOSING", "CLOSED"];
...
+  "passphrase",
+  "servername",
+  "serverName",
+  "ciphers",
+  "dhParamsFile",
+  "secureOptions",
+  "keyFile",
+  "certFile",
+  "caFile",
+  "ALPNProtocols",
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Confirmed and fixed in d1a3320. A top-level `ALPNProtocols: ["http/1.1"]` did throw `TypeError: TLSOptions.ALPNProtocols must be of type string, ArrayBuffer, or null` at construction. Dropped `ALPNProtocols` from `tlsValueKeys` (the agent path already excluded it for the same reason), so it goes back to a silent no-op. Added a test that a string[] ALPNProtocols connects instead of throwing (fails against the previous commit, passes now).

File: src/js/thirdparty/ws.js [hunk 4]
```diff
@@ -170,8 +198,13 @@ class BunWebSocket extends EventEmitter {
...
-        if (!tlsOptions && agentOpts.tls) {
-          tlsOptions = agentOpts.tls;
+        // Merge rather than replace: in `ws`, top-level TLS options and the
+        // agent's connect options both reach the connection. Keep top-level
+        // keys winning on conflict while letting the agent fill the gaps (e.g.
+        // an HttpsProxyAgent carrying cert/key alongside a top-level
+        // rejectUnauthorized), instead of dropping one side wholesale.
+        if (agentOpts.tls) {
+          tlsOptions = tlsOptions ? { ...agentOpts.tls, ...tlsOptions } : agentOpts.tls;
         }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Confirmed and fixed in 3aaf125. Verified the leak: with an explicit `tls: {}` + agent `{ connectOpts: { rejectUnauthorized: false } }`, the agent value leaked in and the self-signed server connected (verification disabled). Guarded the merge on `!explicitTls`, so an explicit `tls` object is a true hard override again while the merge still applies to the extracted top-level path. Added a test asserting the agent's rejectUnauthorized:false does not leak into an explicit tls object (fails before this commit, passes now).

File: src/js/thirdparty/ws.js [hunk 5]
```diff
@@ -158,7 +187,11 @@ class BunWebSocket extends EventEmitter {
     if ($isObject(options)) {
       headers = options?.headers;
       proxy = options?.proxy;
-      tlsOptions = options?.tls;
+      // `ws` forwards top-level TLS options (rejectUnauthorized, ca, cert, ...)
+      // to https.request/tls.connect. An explicit `tls` object stays
+      // authoritative for back-compat with Bun's existing shape.
+      const explicitTls = $isObject(options.tls);
+      tlsOptions = explicitTls ? options.tls : extractTlsOptions(options);
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js:264]: Addressed in 73e3cff, with a tweak. Applying `!= null` to all value keys regressed the file/identity keys: an empty-string `ca: ""` would then be forwarded, and the native parser treats it as a real (empty) CA set that breaks the handshake even with `rejectUnauthorized: false` (verified). So I split the set — scalar keys (`passphrase`, `secureOptions`, `clientRenegotiationLimit`, `clientRenegotiationWindow`) now forward on `!= null` to preserve explicit falsy values like `passphrase: ""`, matching `options.tls`; file/identity keys (`ca`, `cert`, `key`, `caFile`, `servername`, `ciphers`, ...) stay on a truthy check so an empty string still means absent. Note `secureOptions`/`clientRenegotiation*` default to 0 in SSLConfig so forwarding 0 is a no-op, but the paths are now consistent.

File: src/js/thirdparty/ws.js [hunk 6]
```diff
@@ -16,46 +16,75 @@ const readyStates = ["CONNECTING", "OPEN", "CLOSING", "CLOSED"];
...
+  "serverName",
+  "ciphers",
+  "dhParamsFile",
+  "secureOptions",
+  "keyFile",
+  "certFile",
+  "caFile",
+  "clientRenegotiationLimit",
+  "clientRenegotiationWindow",
+];
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Good catch — confirmed a top-level key:[{pem,passphrase}] threw TypeError: element of TLSOptions.key must be of type string, ArrayBuffer, or Blob at construction. Fixed in b390313 with the object-array-guard approach you suggested: key/cert can't be dropped (the string/Buffer forms are the core of this PR), so `isForwardableFileValue` skips only the object-array shape — it falls back to the pre-PR no-op instead of throwing, while string/Buffer/string[]/Blob still forward (verified all connect). I didn't normalize {pem}->pem since SSLConfig.fromJS can't represent per-key passphrases anyway, so honoring that form would be misleading. Added a test that an object-array key constructs without throwing (fails before this commit, passes now).

File: src/js/thirdparty/ws.js [hunk 7]
```diff
@@ -16,46 +16,114 @@ const readyStates = ["CONNECTING", "OPEN", "CLOSING", "CLOSED"];
...
+        !ArrayBuffer.isView(element) &&
+        !(element instanceof Blob) &&
+        !(element instanceof ArrayBuffer)
+      ) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js:92]: Good catch — confirmed a bare `key: { pem, passphrase }` threw TypeError: TLSOptions.key must be of type string, ArrayBuffer, Blob... while the array-wrapped form was skipped, exactly the asymmetry you describe. Fixed in e57956c: factored the shape check into `isRepresentableFileElement` and applied it to the bare value too, so both the wrapped and unwrapped object forms stay a no-op instead of throwing. Verified string/string[]/array-object/bare-object all behave consistently now, and added a test for the bare form (fails before this commit, passes after).

**Comment by Luligu** [comment]
Hello,
any news on this fix ?

Any code for Node.js running with bun runtime and using wss silently fails!

**Comment by robobun** [comment]
Status: the fix is complete and tested. The shim now forwards the top-level TLS options (`rejectUnauthorized`, `ca`, `cert`, `key`, ...) to the connection, with seven tests covering the main path and the edge cases from review. All review feedback is addressed and the full test suite passed on every CI platform on the last run. The remaining step is a maintainer sign-off, since this touches TLS verification semantics.

`main` has moved since the last push, so I am rebasing the branch now to keep it mergeable.

Until it lands, two workarounds exist on current Bun releases:

1. Pass the options under Bun's `tls` key, which the shim already honors:
   ```js
   new WebSocket("wss://localhost:8443", { tls: { rejectUnauthorized: false } });
   ```
2. Or set `NODE_TLS_REJECT_UNAUTHORIZED=0` in the environment (disables verification process-wide, so prefer option 1).

**Comment by Luligu** [comment]
> Status: the fix is complete and tested. The shim now forwards the top-level TLS options (`rejectUnauthorized`, `ca`, `cert`, `key`, ...) to the connection, with seven tests covering the main path and the edge cases from review. All review feedback is addressed and the full test suite passed on every CI platform on the last run. The remaining step is a maintainer sign-off, since this touches TLS verification semantics.
> 
> `main` has moved since the last push, so I am rebasing the branch now to keep it mergeable.
> 
> Until it lands, two workarounds exist on current Bun releases:
> 
> 1. Pass the options under Bun's `tls` key, which the shim already honors:
>    ```js
>    new WebSocket("wss://localhost:8443", { tls: { rejectUnauthorized: false } });
>    ```
> 2. Or set `NODE_TLS_REJECT_UNAUTHORIZED=0` in the environment (disables verification process-wide, so prefer option 1).

I'm well aware of this trick thanks.

Hope this will be fixed asap.

File: src/js/thirdparty/ws.js [hunk 8]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+// `ws` forwards top-level TLS options to `https.request`/`tls.connect`; Bun
+// reads them from the `tls` option, so collect them off the top-level options.
+// Booleans use `!== undefined` so an explicit `false` (e.g. `rejectUnauthorized:
+// false`, the whole point of connecting to a self-signed server) is forwarded.
+//
+// `ALPNProtocols` is intentionally omitted: Node/`ws` accept the `string[]`
+// form, but SSLConfig.fromJS only takes string/ArrayBuffer/null and throws on
+// an array. Forwarding it would turn a previously-ignored option into a
+// constructor throw, so leave it a no-op (WebSocket negotiates subprotocols via
+// Sec-WebSocket-Protocol, not TLS ALPN).
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 9]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+const tlsBooleanKeys = ["rejectUnauthorized", "requestCert", "lowMemoryMode"];
+// Scalar options where a falsy value is a legitimate setting (0, ""), so they
+// are forwarded on `!= null` rather than truthiness — matching how `options.tls`
+// would pass them through. `secureOptions`/`clientRenegotiation*` default to 0
+// in SSLConfig, so 0 is effectively a no-op, but this keeps the two paths
+// consistent and preserves an explicit `passphrase: ""`.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 10]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+const tlsScalarKeys = ["passphrase", "secureOptions", "clientRenegotiationLimit", "clientRenegotiationWindow"];
+// File/identity options where an empty string means "absent": an empty `ca`
+// (etc.) must not be forwarded, or the native parser treats it as a real (empty)
+// value and the handshake breaks. These stay on a truthy check.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 11]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+  "caFile",
+  "servername",
+  "serverName",
+  "ciphers",
+];
+const tlsValueKeys = [...tlsScalarKeys, ...tlsFileKeys];
+
+// Agents (e.g. HttpsProxyAgent) stuff connection options into `connectOpts`
+// that aren't all valid Bun TLS options, so keep the agent path to the subset
+// ws users historically pass through an agent.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 12]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+/**
+ * A file/identity value SSLConfig.fromJS can parse: string | ArrayBuffer | Blob,
+ * or an array of those. Node/`ws` also accept the `key`/`cert` object form
+ * (`{ pem, passphrase }`, bare or in an array) for per-key passphrases, but the
+ * native parser has no arm for a plain object and throws. Skip those shapes so
+ * they stay a no-op (as they were before top-level TLS options were forwarded)
+ * instead of becoming a throw.
+ * @param {unknown} value
+ * @returns {boolean}
+ */
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 13]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+  if (!value) return false;
+  if (Array.isArray(value)) {
+    for (const element of value) {
+      if (!isRepresentableFileElement(element)) return false;
+    }
+    return true;
+  }
+  // A bare value must be one of the same representable shapes; a plain
+  // `{ pem, passphrase }` object (or a KeyObject) has no SSLConfig arm, so skip
+  // it like its array-wrapped form rather than forwarding it into a throw.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 14]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+  return isRepresentableFileElement(value);
+}
+
+/**
+ * Whether a single file/identity value is a shape SSLConfig.fromJS accepts:
+ * string, ArrayBuffer, TypedArray/DataView, or Blob. A plain object is not.
+ * @param {unknown} value
+ * @returns {boolean}
+ */
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 15]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+
+/**
+ * Collects the TLS options the native WebSocket understands from a source
+ * object. `valueKeys` narrows the set for callers that can't pass the full set
+ * (see {@link agentTlsValueKeys}). Scalar keys (see {@link tlsScalarKeys})
+ * preserve explicit falsy values; everything else is copied when truthy.
+ * @param {Object} source
+ * @param {string[]} [valueKeys]
+ * @returns {Object|null} The TLS options, or null if none were present
+ */
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 16]
```diff
@@ -30,47 +30,120 @@ function sendAfterClose(state, cb) {
...
+  let tls = null;
+  for (const key of tlsBooleanKeys) {
+    if (source[key] !== undefined) {
+      (tls ??= {})[key] = source[key];
+    }
+  }
+  for (const key of valueKeys) {
+    // Scalars forward an explicit falsy value (0, "") like `options.tls` would;
+    // file/identity keys treat a falsy value (e.g. `ca: ""`) as absent and skip
+    // the object-array form the native parser can't represent.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 17]
```diff
@@ -202,7 +275,11 @@ class BunWebSocket extends EventEmitter {
     if ($isObject(options)) {
       headers = options?.headers;
       proxy = options?.proxy;
-      tlsOptions = options?.tls;
+      // `ws` forwards top-level TLS options (rejectUnauthorized, ca, cert, ...)
+      // to https.request/tls.connect. An explicit `tls` object stays
+      // authoritative for back-compat with Bun's existing shape.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 18]
```diff
@@ -215,8 +292,13 @@ class BunWebSocket extends EventEmitter {
         if (!proxy && agentProxy) {
           proxy = agentProxy;
         }
-        if (!tlsOptions && agentTls) {
-          tlsOptions = agentTls;
+        // An explicit `tls` object is a hard override — don't let an agent's
+        // connect options (which semantically target the proxy hop) leak into
+        // it. Otherwise merge: top-level keys win on conflict while the agent
+        // fills gaps (e.g. an HttpsProxyAgent carrying cert/key alongside a
+        // top-level rejectUnauthorized), instead of dropping one side wholesale.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 19]
```diff
@@ -30,47 +30,117 @@ function sendAfterClose(state, cb) {
...
+const ObjectPrototypeHasOwnProperty = Object.prototype.hasOwnProperty;
+const ArrayBufferIsView = ArrayBuffer.isView;
+const NativeArrayBuffer = ArrayBuffer;
+const NativeBlob = Blob;
+
+// How each top-level TLS option ws forwards to tls.connect is copied into the
+// native `tls` option. Limited to the keys SSLConfig.fromJS parses.
+// ALPNProtocols is omitted: ws takes a string[] and SSLConfig.fromJS throws on
+// an array. minVersion/maxVersion are omitted: ws takes a string, SSLConfig
+// takes a number.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 20]
```diff
@@ -30,47 +30,117 @@ function sendAfterClose(state, cb) {
...
+  servername: TLS_FILE,
+  serverName: TLS_FILE,
+  ciphers: TLS_FILE,
+  sigalgs: TLS_FILE,
+  ecdhCurve: TLS_FILE,
+};
+const tlsKeys = Object.keys(tlsKeyKinds);
+
+// HttpsProxyAgent.connectOpts carries options SSLConfig.fromJS rejects, so
+// the agent path forwards only these.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 21]
```diff
@@ -30,47 +30,117 @@ function sendAfterClose(state, cb) {
...
+const agentTlsKeys = ["rejectUnauthorized", "ca", "cert", "key", "passphrase"];
+
+// SSLConfig.fromJS accepts string | ArrayBuffer | Blob, or an array of those.
+// ws also accepts `{ pem, passphrase }` (bare or in an array) for key/cert;
+// that shape is skipped because the native parser throws on a plain object.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 22]
```diff
@@ -30,47 +30,117 @@ function sendAfterClose(state, cb) {
...
+  );
+}
+
+/**
+ * Reads own properties only: a polluted `Object.prototype.rejectUnauthorized`
+ * must not disable verification for every socket.
+ * @param {Object} source
+ * @param {string[]} [keys]
+ * @returns {Object|null}
+ */
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 23]
```diff
@@ -215,8 +287,11 @@ class BunWebSocket extends EventEmitter {
         if (!proxy && agentProxy) {
           proxy = agentProxy;
         }
-        if (!tlsOptions && agentTls) {
-          tlsOptions = agentTls;
+        // Agent TLS fills gaps under the top-level keys. It never reaches an
+        // explicit `tls` object: the agent's rejectUnauthorized targets the
+        // proxy hop and must not weaken verification of the target.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: src/js/thirdparty/ws.js [hunk 24]
```diff
@@ -30,47 +30,105 @@ function sendAfterClose(state, cb) {
...
 }
 
+// Captured at load so a user script cannot redirect the TLS option checks.
+const ObjectPrototypeHasOwnProperty = Object.prototype.hasOwnProperty;
+const ArrayBufferIsView = ArrayBuffer.isView;
+const NativeArrayBuffer = ArrayBuffer;
+const NativeBlob = Blob;
+
+// Top-level TLS keys SSLConfig.fromJS parses. ALPNProtocols and min/maxVersion
+// are left out because ws and SSLConfig disagree on their type.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/js/thirdparty/ws.js]: Trimmed in 0ca973b: each comment is now one or two lines of fact (what the key set is, which shapes SSLConfig.fromJS rejects), with the justification prose removed.

File: test/js/first_party/ws/ws-tls-options.test.ts [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+  // (Array.isArray is left alone: the pre-existing protocol parsing at the top
+  // of the constructor still calls it.)
+  it("still forwards options after Array.prototype.includes and ArrayBuffer.isView are clobbered", async () => {
+    const out = await runInChild(`${childPreamble}
+      Array.prototype.includes = () => { throw new Error("clobbered"); };
+      ArrayBuffer.isView = () => { throw new Error("clobbered"); };
+      console.log(await connect({ rejectUnauthorized: false, key: [tls.key], cert: tls.cert }));
+      server.stop(true);
+    `);
+    expect(out).toBe("open");
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/first_party/ws/ws-tls-options.test.ts]: Good catch, the test was vacuous: a PEM string short-circuits at !$isObject before the captured ArrayBufferIsView is ever reached, and .includes is gone from this path. Fixed in 0ca973b: the child now passes key: [Buffer.from(tls.key)] and cert: Buffer.from(tls.cert), so the file-shape check really invokes the captured ArrayBufferIsView, and the dead Array.prototype.includes clobber is dropped. Verified it is load-bearing: with the capture reverted to a live ArrayBuffer.isView(value) lookup the test fails (error: clobbered); with the capture it passes. Checked twice.
